### PR TITLE
clk based PWM

### PR DESCRIPTION
**Pull request by TravMurav**
Opening to make the review easier

This contains 3 series:

 - msm8916: Prepare general purpose clocks
 - Add pwm-clk
 - Extend sgm3140 led driver
 
 TODO:

- [ ] pinctrl-msm8916: Rename that strobe to proper GP_CLK{0,1} conforming to the docs
- [ ] DT schema everything
- [ ] remove all debug lines and useless comments

sent upstream:
* https://lists.sr.ht/~postmarketos/upstreaming/patches/27235
* https://lists.sr.ht/~postmarketos/upstreaming/patches/27237

### CODE REVIEW

**Comment by minlexx** [comment]
```
(Doesn't seem to bedescribed in gcc,
    ignored in this patch)
```
=>
```
(Doesn't seem to be described in gcc,
    ignored in this patch)
```
typo in commit 557128285c90a0ef128cace67df33eb4f59d5893 desc?..

**Comment by TravMurav** [comment]
> typo in commit [5571282](https://github.com/msm8916-mainline/linux/commit/557128285c90a0ef128cace67df33eb4f59d5893) desc?..

Oh, no...

```
LANG=C git st
On branch pwm-clk
Your branch and 'msm8916/pwm-clk' have diverged,
and have 5 and 5 different commits each, respectively.
```

I fixed the typo but forgot to push it :D
Thanks for noticing

File: drivers/pinctrl/qcom/pinctrl-msm8916.c [hunk 1]
```diff
@@ -844,8 +844,8 @@ static const struct msm_pingroup msm8916_groups[] = {
 	PINGROUP(28, pwr_modem_enabled_a, NA, NA, NA, NA, NA, qdss_tracedata_b, NA, atest_combodac),
 	PINGROUP(29, cci_i2c, NA, NA, NA, NA, NA, qdss_tracedata_b, NA, atest_combodac),
 	PINGROUP(30, cci_i2c, NA, NA, NA, NA, NA, NA, NA, qdss_tracedata_b),
-	PINGROUP(31, cci_timer0, NA, NA, NA, NA, NA, NA, NA, NA),
-	PINGROUP(32, cci_timer1, NA, NA, NA, NA, NA, NA, NA, NA),
+	PINGROUP(31, cci_timer0, flash_strobe, NA, NA, NA, NA, NA, NA, NA),
+	PINGROUP(32, cci_timer1, flash_strobe, NA, NA, NA, NA, NA, NA, NA),
```
Comment (Minecrell) [inline drivers/pinctrl/qcom/pinctrl-msm8916.c:848]: The naming is a bit weird here. What's the difference compared to the `gcc_gp1_clk_a` function?

File: drivers/pwm/pwm-clk.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		return 0;
+	}
+
+	rate = div64_u64(NSEC_PER_SEC, state->period);
+	ret = clk_set_rate(chip->clk, rate);
+	if (ret)
+		return ret;
+
+	ret = clk_set_duty_cycle(chip->clk, state->duty_cycle, state->period);
+	return ret;
```
Comment (Minecrell) [inline drivers/pwm/pwm-clk.c:45]: return `clk_set_duty_cycle()`?

File: drivers/leds/flash/leds-sgm3140.c [hunk 3]
```diff
@@ -35,38 +47,86 @@ static struct sgm3140 *flcdev_to_sgm3140(struct led_classdev_flash *flcdev)
...
+	return 0;
+}
+
 static int sgm3140_strobe_set(struct led_classdev_flash *fled_cdev, bool state)
 {
 	struct sgm3140 *priv = flcdev_to_sgm3140(fled_cdev);
 	int ret;
 
-	if (priv->enabled == state)
+	//FIXME it ignores strobe if torch mode was enabled?
```
Comment (Minecrell) [inline drivers/leds/flash/leds-sgm3140.c:112]: fixme

**Comment by stephan-gh** [comment]
I think there is some pwm-clk stuff in rebase now, can you make a new PR for the dts changes if needed?

**Pull request closed without merging**